Templates/Templates/AutoDecodable.stencil

```diff
@@ -9,28 +11,85 @@ extension KeyedDecodingContainer {
+  {% set encodeCall %}{% if var.isOptional %}encodeIfPresent{% else %}encode{% endif %}{% endset %}
+  {% if var|!annotated:"encodingType" %}
+    try container.{{ encodeCall }}({{ var.name }}, forKey: .{{ var.name }})
+  {% else %}
+  {% set typename %}{{ var.typeName.name }}{% endset %}
+  {% set encodingType %}{{ var.annotations.encodingType }}{% endset %}
+  {% if typename == "Bool" and encodingType == "String" %}
+    try container.encode({{ var.name }} ? 'true' : 'false', forKey: .{{ var.name }})
+  {% elif typename == "Bool" and encodingType == "Int" %}
+    try container.encode({{ var.name }} ? 1 : 0, forKey: .{{ var.name }})
+  {% endif %}
+  {% endif %}
+{% endmacro %}
+
+{% macro encode type %}
+  {{ type.accessLevel }} {% if type.supertype.implements.Encodable or type.supertype.implements.AutoEncodable %}override {% endif %}func encode(to encoder: Encoder) throws {
+    var container = encoder.container(keyedBy: CodableKeys.self)
+    {% for variable in type.instanceVariables where variable|!annotated:"skipCoding" %}
+    {% call convertTypeForEncoder variable %}
     {% endfor %}
   }
 {% endmacro %}
 
-{# going to exclude protocols for now, handle it seperatly #}
-{% for type in types.all|!protocol where type.based.AutoDecodable or type|annotated:"AutoDecodable" and not type.name == "AutoCodable" and not type.name == "AutoDecodable" %}
-  // MARK: {{ type.name }}+Decodable | AutoDecodable
+{% for type in types.all|!class where type.based.AutoDecodable or type|annotated:"AutoDecodable" and type.name != "AutoCodable" and type.name != "AutoDecodable" %}
+// MARK: {{ type.name }}+Decodable | AutoDecodable
+extension {{ type.name }} {
+  {% call codingKeys type %}
+
+  {{ type.accessLevel }} init(from decoder: Decoder) throws {
+    let container = try decoder.container(keyedBy: CodableKeys.self)
+    {% for var in type.instanceVariables where var|!annotated:"skipCoding" %}
+    {% if var|annotated:"codingDefaultValue" or var.isOptional %}
+    {{ var.name }} = try container.decodeOrDefault({{ var.typeName.name }}.self, forKey: .{{ var.name }}, defaultValue: {% call defaultValueFor var %}
+    {% else %}
+    {{ var.name }} = try container.decode({{ var.typeName.name }}.self, forKey: .{{ var.name }})
+    {% endif %}
+    {% endfor %}
+  }
+}
+{% endfor %}
+
+{# todo inheritance #}
+{% for type in types.all|class where type.based.AutoDecodable or type|annotated:"AutoDecodable" and type.name != "AutoCodable" and type.name != "AutoDecodable" %}
   // sourcery:inline:auto:{{ type.name }}.AutoDecodable
   {% call codingKeys type %}
-  {{ type.accessLevel }}{% if type|class %} convenience required{% endif %} init(from decoder: Decoder) throws {
-    {% if type|class %}self.init(){% endif %}
+  {% if type.supertype.based.AutoEncodable or type.based.AutoEncodable %}
+  {% call encode type %}
+  {% endif %}
+  {{ type.accessLevel }} convenience required init(from decoder: Decoder) throws {
+    {% call inheritanceInitializer type %}
```

Comment:
why is this constructor marked as convenience and why this macro is called at the beginning? Shouldn't it just call `super.init` in the end? Also isn't it that only own properties of subclass should be decoded (and have keys defined for them, though extra keys will not raise issues) and the rest should be delegated to super class constructor?